File: lab3b.py
import random
import math

def crie_matriz_aleatoria(n_linhas, n_colunas):
    
    matriz = []    
    
    for i in range(int(n_linhas)):    
        linha = []        
        for j in range(int(n_linhas)):        
            linha.append(random.randint(1,256))           
    
    matriz.append(linha)    
    
    return matriz

def matriz_GLCM(M, n_linhas, n_colunas, N):
    
    M = list()         
    GLCM = list()
    
    for i in range(int(N)):    
        linha = []        
        for j in range(int(N)):        
            linha.append(0)           
        GLCM.append(linha)

    for i in range(int(n_linhas)):    
        for j in range(int(n_colunas)-1):        
            a = M[i][j]
            b = M[i][j+1]
            GLCM[a][b] +=1
            
    return GLCM

def dimGLCM (M, n_linhas, n_colunas):
    A = list()
    N = 0
    
    for i in range(int(n_linhas)):
        linha = []
        for j in range(int(n_colunas)):  
           linha.append(M[i][j])           
        A.append(linha) 
            
            
    for i in range(int(n_linhas)):        
         for j in range(int(n_colunas)):        
             if ( A[i][j] > N):
                 N = A[i][j] + 1
    return N

print("Digite o numero de linhas da matriz:\n")
linhas = input()

print("Digite o numero de colunas da matriz:\n")
colunas = input()

mat = crie_matriz_aleatoria(linhas, colunas)

#dimGLCM = 0    

Nglcm = dimGLCM(mat, linhas, colunas)
             
G =  matriz_GLCM(mat, linhas, colunas, Nglcm)

for i in range(linhas):
    for j in range(colunas):
            print(mat[i][j])
    print("\n")
    
for i in range(Nglcm):    
    for j in range(Nglcm):        
            print(G[i][j])
    print("\n")     
    
energia = 0
entropia = 0
contraste = 0
homogeneidade = 0

for i in range(Nglcm):    
    for j in range(Nglcm):        
            energia += (G[i][j])*(G[i][j])
    print("\n")

for i in range(Nglcm):    
    for j in range(Nglcm):        
            entropia += -math.log(G[i][j])*(G[i][j])
    print("\n")     

for i in range(Nglcm):    
    for j in range(Nglcm):        
            contraste += (G[i][j])*(i-j)*(i-j)
    print("\n")
    
for i in range(Nglcm):    
    for j in range(Nglcm):        
            homogeneidade += (G[i][j])/(1+(i-j)*(i-j))
    print("\n")

print(energia)
print(entropia)
print(contraste)
print(homogeneidade)

File: lab3b.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

void matrizGLCM (int *, int, int);

int main()
{
    int i, j;
    int lin, col, *mat;       // Dimensoes matriz original

    printf("Digite o numero de linhas da matriz:\n");
    scanf("%d",&lin);

    printf("Digite o numero de colunas da matriz:\n");
    scanf("%d",&col);

    srand(time(NULL));

    mat = (int*)malloc(lin*col*sizeof(int));

    for (i=0;i<lin;i++){
        for (j=0;j<col;j++){
            *(mat+col*i+j) = rand() %8;
        }
    }

    
    printf("\nMATRIZ: \n\n");
        for (i=0;i<lin;i++){
            for (j=0;j<col;j++)
        {
            printf("%d \t", *(mat+col*i+j));
        }
        printf("\n");
    }

    matrizGLCM(mat, lin, col);

    free(mat);

    return 0;
}

void matrizGLCM (int *M, int LIN, int COL){
    int i, j;
    int a, b;
    int N = 0;
    int *G;
    float energia = 0;
    float entropia = 0;
    float contraste = 0;
    float homogeneidade = 0;

    for (i=0;i<LIN; i++){
        for (j=0;j<COL;j++){
            if(*(M+COL*i+j) > N){
                N = *(M+COL*i+j);
            }
        }
    }

    G = (int*)malloc((N+1)*(N+1)*sizeof(int));

    for (i=0;i<LIN; i++){
        for (j=0;j<COL;j++){
            *(G+(N+1)*i+j) = 0;
        }
    }

    for (i=0;i<LIN;i++){
        for (j=0;j<COL-1;j++){
            a = *(M+COL*i+j);
            b = *(M+COL*i+j+1);
            *(G+(N+1)*a+b)+=1;
        }
    }

    printf("\nGLCM: \n\n");
        for (i=0;i<LIN;i++){
            for (j=0;j<COL;j++)
        {
            printf("%d \t", *(G+(N+1)*i+j));
        }
        printf("\n");
    }

    for (i=0;i<LIN; i++){
        for (j=0;j<COL;j++){
            energia += (*(G+(N+1)*i+j))*(*(G+(N+1)*i+j));
        }
    }

    for (i=0;i<LIN; i++){
        for (j=0;j<COL;j++){
            entropia += -log(*(G+(N+1)*i+j))*(*(G+(N+1)*i+j));
        }
    }

    for (i=0;i<LIN; i++){
        for (j=0;j<COL;j++){
            contraste += (*(G+(N+1)*i+j))*(i-j)*(i-j);
        }
    }

      for (i=0;i<LIN; i++){
        for (j=0;j<COL;j++){
            homogeneidade += (*(G+(N+1)*i+j))/(1+(i-j)*(i-j));
        }
    }

    printf("\nenergia: \n", energia);
    printf("\nentropia: \n", entropia);
    printf("\ncontraste: \n", contraste);
    printf("\nhomogeneidade: \n",homogeneidade);

}  